File: java/src/kanzi/function/NullFunction.java
/*
Copyright 2011, 2012 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.function;

import kanzi.ByteFunction;
import kanzi.IndexedByteArray;


public class NullFunction implements ByteFunction
{
   private final int size;


   public NullFunction()
   {
      this(0);
   }


   public NullFunction(int size)
   {
      if (size < 0)
         throw new IllegalArgumentException("Invalid size parameter (must be at least 0)");

      this.size = size;
   }


   public int size()
   {
      return this.size;
   }

   
   @Override
   public boolean forward(IndexedByteArray source, IndexedByteArray destination)
   {
      return doCopy(source, destination, this.size);
   }
  

   @Override
   public boolean inverse(IndexedByteArray source, IndexedByteArray destination)
   {    
      return doCopy(source, destination, this.size);
   }

   
   private static boolean doCopy(IndexedByteArray source, IndexedByteArray destination, int sz)
   {
      if ((source.array == destination.array) && (source.index == destination.index))
         return true;
      
      final int len = (sz == 0) ? source.array.length : sz;

      if (source.index + len > source.array.length)
         return false;
      
      if (destination.index + len > destination.array.length)
         return false;

      System.arraycopy(source.array, source.index, destination.array, destination.index, len);
      return true;
   }
}

File: go/src/kanzi/function/NullFunction.go
/*
Copyright 2011, 2012 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package function

import (
	"errors"
)

type NullFunction struct {
	size uint
}

func NewNullFunction(sz uint) (*NullFunction, error) {
	this := new(NullFunction)
	this.size = sz
	return this, nil
}

func (this *NullFunction) Size() uint {
	return this.size
}

func doCopy(src, dst []byte, sz uint) (uint, uint, error) {
	length := sz

	if sz == 0 {
		length = uint(len(src))
	}

	if length > uint(len(dst)) {
		return uint(0), uint(0), errors.New("Destination buffer too small")
	}

	copy(dst, src[0:length])
	return length, length, nil
}

func (this *NullFunction) Forward(src, dst []byte) (uint, uint, error) {
	return doCopy(src, dst, this.size)
}

func (this *NullFunction) Inverse(src, dst []byte) (uint, uint, error) {
	return doCopy(src, dst, this.size)
}
